### PR TITLE
WIP: New module: GitLab service

### PR DESCRIPTION
**Pull request by drzraf**
##### SUMMARY
New module: **Gitlab integration services**

Fixes https://github.com/ansible/ansible/pull/40053

##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
gitlab_service

##### ADDITIONAL INFORMATION
This module provides a way to setup Gitlab integration services using Ansible and supports all services defined by Gitlab.com API (and their corresponding parameters) as of 2020/07/09.

Known bugs:
- Most Gitlab integration services rely upon _secret_ parameter (password, token, api_key) which are write-only (ie: not exposed on GET request). That currently [keeps module from being idempotent when such a suboption is used](https://gitlab.com/gitlab-org/gitlab-ce/issues/46313).

Needs work:
- The _active_ state of a service [could not be set using the API](https://gitlab.com/gitlab-org/gitlab-ce/issues/41113). Use `state=absent` to completely remove the service configuration instead of deactivating it. But some work happened recently on GitLab side, it may be fixed by now.

##### EXTENDED DEFINITION OF SERVICE
- Their validation is done by doing a redundant initialization of the `AnsibleModule()` and modifying the argument spec meanwhile.

- The extended definition of services and their suboptions (**name**, **required** or not, **type**) is generated from [Gitlab upstream file](https://gitlab.com/gitlab-org/gitlab/-/blob/master/lib/api/helpers/services_helpers.rb) the 2020/07/09.

### CODE REVIEW

**Comment by drzraf** [comment]
As stated in the original task I won't go through the horrible hassle to get a Python 2.6 toolchain for the sole purpose of lint'ing a module that will never be realistically used with such a version.

**Comment by felixfontein** [comment]
There's no need to install a 2.6 toolchain. You can either run `ansible-test sanity --docker -v`, which includes among others a Python 2.6 in the container, or you can use CI. In any case, if the module does not conform to Python 2.6 syntax, it will not be merged.

In any case, the module must not live in `plugins/modules/`, but in a subdirectory. You need to create a symlink from `plugins/modules/` to the actual module.

**Comment by Andersson007** [comment]
@drzraf any updates on this?
we expanded gitlab_team a little bit, so i hope, there will be people to review the module. If not, i think, community team will review this

**Comment by drzraf** [comment]
Screenshot from 2020-10-22 20-10-17

...

Then I tried to get the [good old local testing](https://docs.ansible.com/ansible/latest/dev_guide/developing_modules_general.html#exercising-module-code-locally) (`python -m ansible.modules.my_test /tmp/args.json`), but I don't see how (_if?_) it's possible nowadays.

File: changelogs/fragments/667-gitlab_service.yaml [hunk 1]
```diff
@@ -0,0 +1,2 @@
+minor_changes:
+- gitlab_service - new module to setup GitLab integration services (https://github.com/ansible-collections/community.general/pull/667)
```
Comment (felixfontein) [inline changelogs/fragments/667-gitlab_service.yaml:2]: Please remove the changelog fragment. New modules must not have one.

File: plugins/modules/source_control/gitlab/gitlab_service.py [hunk 2]
```diff
@@ -0,0 +1,467 @@
...
+requirements:
+  - python >= 2.6
+  - python-gitlab python module
+extends_documentation_fragment:
+- community.general.auth_basic
+
+options:
+  project:
+    description:
+      - The ID or path of the project (urlencoded or not)
```
Comment (felixfontein) [inline plugins/modules/source_control/gitlab/gitlab_service.py:28]:
```suggestion
      - The ID or path of the project (urlencoded or not).
```

File: plugins/modules/source_control/gitlab/gitlab_service.py [hunk 3]
```diff
@@ -0,0 +1,467 @@
...
+    required: true
+    type: str
+  active:
+    description:
+      - Not yet supported (U(https://gitlab.com/gitlab-org/gitlab-ce/issues/41113))
```
Comment (felixfontein) [inline plugins/modules/source_control/gitlab/gitlab_service.py:33]:
```suggestion
      - Not yet supported (U(https://gitlab.com/gitlab-org/gitlab-ce/issues/41113)).
```
Always make paragraphs (list elements) in description proper sentences. I won't comment for the others.

File: plugins/modules/source_control/gitlab/gitlab_service.py [hunk 4]
```diff
@@ -0,0 +1,467 @@
...
+  params:
+    description:
+      - The description of the service, see documentation U(https://docs.gitlab.com/ee/api/services.html)
+    type: dict
+  events:
+    description:
+      - The events that trigger the service (required if C(state=present))
+    choices: ["push", "issues", "confidential_issues", "merge_requests", "tag_push", "note", "confidential_note", "job", "pipeline", "wiki_page"]
+    default: ["push", "issues", "confidential_issues", "merge_requests", "tag_push", "note", "confidential_note", "job", "pipeline", "wiki_page"]
+    type: list
```
Comment (felixfontein) [inline plugins/modules/source_control/gitlab/gitlab_service.py:83]:
```suggestion
    type: list
    elements: str
```

File: plugins/modules/source_control/gitlab/gitlab_service.py [hunk 5]
```diff
@@ -0,0 +1,467 @@
...
+  state:
+    description:
+      - Create or delete a service.
+      - Possible values are present and absent.
```
Comment (felixfontein) [inline plugins/modules/source_control/gitlab/gitlab_service.py:87]:
```suggestion
      - Possible values are C(present) and C(absent).
```

File: plugins/modules/source_control/gitlab/gitlab_service.py [hunk 6]
```diff
@@ -0,0 +1,467 @@
...
+    default: "present"
+    choices: ["present", "absent"]
+    type: str
+'''
+
+EXAMPLES = '''
+# Setup email on push for this project
+- name: emails me on push
+  gitlab_service:
```
Comment (felixfontein) [inline plugins/modules/source_control/gitlab/gitlab_service.py:96]:
```suggestion
  community.general.gitlab_service:
```
Always use FQCN in examples.

File: plugins/modules/source_control/gitlab/gitlab_service.py [hunk 7]
```diff
@@ -0,0 +1,467 @@
...
+  delegate_to: localhost
+
+- Idempotency is only partially provided since GitLab does
+  not expose secret parameters like tokens or passwords.
+  See U(https://gitlab.com/gitlab-org/gitlab-ce/issues/46313)
+'''
+
+RETURN = '''
+---
+diff:
```
Comment (felixfontein) [inline plugins/modules/source_control/gitlab/gitlab_service.py:128]: Standard return values do not need to be documented. Simply state somewhere that you support diff mode.

**Review by felixfontein — COMMENTED** [review]
Please note that all new modules must come with tests, i.e. unit and/or integration tests.

File: plugins/modules/source_control/gitlab/gitlab_service.py [hunk 8]
```diff
@@ -0,0 +1,467 @@
...
+# Copyright: (c) 2020, Raphaël Droz <raphael@droz.eu>
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+from __future__ import absolute_import, division, print_function
+__metaclass__ = type
+
+DOCUMENTATION = '''
+---
+module: gitlab_service
+short_description: Creates, updates or deletes GitLab services (aka "integrations").
```
Comment (Andersson007) [inline plugins/modules/source_control/gitlab/gitlab_service.py:13]:
```suggestion
short_description: Creates, updates or deletes GitLab services (aka "integrations")
```
Please, also remove the acronym

File: plugins/modules/source_control/gitlab/gitlab_service.py [hunk 9]
```diff
@@ -0,0 +1,467 @@
...
+version_added: '1.3.0'
+description:
+  - Setup or delete GitLab integration services.
```
Comment (Andersson007) [inline plugins/modules/source_control/gitlab/gitlab_service.py:16]:
```suggestion
  - Setup or delete GitLab integration services.
```
the short description is longer than the description, you could just make them equal (the final dot must be in a `description` and must not be in a `short_description`)

File: plugins/modules/source_control/gitlab/gitlab_service.py [hunk 10]
```diff
@@ -0,0 +1,467 @@
...
+    description:
+      - The ID or path of the project (urlencoded or not)
+    required: true
+    type: str
+  active:
+    description:
+      - Not yet supported (U(https://gitlab.com/gitlab-org/gitlab-ce/issues/41113))
+    type: bool
+    default: yes
+    choices: [true]
```
Comment (Andersson007) [inline plugins/modules/source_control/gitlab/gitlab_service.py:36]:
```suggestion
```

File: plugins/modules/source_control/gitlab/gitlab_service.py [hunk 11]
```diff
@@ -0,0 +1,467 @@
...
+      - unify-circuit
+      - webex-teams
+      - youtrack
+  params:
+    description:
+      - The description of the service, see documentation U(https://docs.gitlab.com/ee/api/services.html)
+    type: dict
+  events:
+    description:
+      - The events that trigger the service (required if C(state=present))
```
Comment (Andersson007) [inline plugins/modules/source_control/gitlab/gitlab_service.py:80]:
```suggestion
      - The events that trigger the service (required if I(state=present))
```

**Comment by felixfontein** [comment]
@drzraf are you still working on this?

needs_info

**Comment by felixfontein** [comment]
Closing since there was no more reaction. @drzraf if you're interested again in working on this, please ping me. If someone else is interested in continuing this PR, please also write it here so we can coordinate.

**Comment by drzraf** [comment]
/ping @felixfontein (Resolved all the comments again)

**Comment by felixfontein** [comment]
@drzraf since you force-pushed, GitHub does not let me open this PR anymore.

**Pull request closed without merging**